test/integration/scheduler/taint_test.go

```diff
@@ -0,0 +1,317 @@
+	internalInformers.WaitForCacheSync(controllerCh)
+
+	// -------------------------------------------
+	// Test TaintNodeByCondition feature.
+	// -------------------------------------------
+	memoryPressureToleration := v1.Toleration{
+		Key:      algorithm.TaintNodeMemoryPressure,
+		Operator: v1.TolerationOpExists,
+		Effect:   v1.TaintEffectNoSchedule,
+	}
+
+	// Case 1: Add MememoryPressure Toleration for non-BestEffort pod.
+	bustablePod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "bustable-pod",
+			Namespace: nsName,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "busybox",
+					Image: "busybox",
+					Resources: v1.ResourceRequirements{
+						Requests: v1.ResourceList{
+							v1.ResourceCPU: resource.MustParse("100m"),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	bustablePodInServ, err := clientset.CoreV1().Pods(nsName).Create(bustablePod)
+	if err != nil {
+		t.Errorf("Case 1: Failed to create pod: %v", err)
+	} else if !reflect.DeepEqual(bustablePodInServ.Spec.Tolerations, []v1.Toleration{memoryPressureToleration}) {
+		t.Errorf("Case 1: Unexpected toleration of non-BestEffort pod, expected: %+v, got: %v",
+			[]v1.Toleration{memoryPressureToleration},
+			bustablePodInServ.Spec.Tolerations)
+	}
+
+	// Case 2: No MemoryPressure Toleration for BestEffort pod.
+	besteffortPod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "best-effort-pod",
+			Namespace: nsName,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "busybox",
+					Image: "busybox",
+				},
+			},
+		},
+	}
+
+	besteffortPodInServ, err := clientset.CoreV1().Pods(nsName).Create(besteffortPod)
+	if err != nil {
+		t.Errorf("Case 2: Failed to create pod: %v", err)
+	} else if len(besteffortPodInServ.Spec.Tolerations) != 0 {
+		t.Errorf("Case 2: Unexpected toleration # of BestEffort pod, expected: 0, got: %v",
+			len(besteffortPodInServ.Spec.Tolerations))
+	}
+
+	// Case 3: Taint Node by NetworkUnavailable condition.
+	networkUnavailableNode := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "node-1",
+		},
+		Status: v1.NodeStatus{
+			Capacity: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("4000m"),
+				v1.ResourceMemory: resource.MustParse("16Gi"),
+				v1.ResourcePods:   resource.MustParse("110"),
+			},
+			Allocatable: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("4000m"),
+				v1.ResourceMemory: resource.MustParse("16Gi"),
+				v1.ResourcePods:   resource.MustParse("110"),
+			},
+			Conditions: []v1.NodeCondition{
+				{
+					Type:   v1.NodeNetworkUnavailable,
+					Status: v1.ConditionTrue,
+				},
+				{
+					Type:   v1.NodeReady,
+					Status: v1.ConditionFalse,
+				},
+			},
+		},
+	}
+
+	if _, err := clientset.CoreV1().Nodes().Create(networkUnavailableNode); err != nil {
+		t.Errorf("Case 3: Failed to create node: %v", err)
+	} else {
+		nodeInformer := informers.Core().V1().Nodes().Informer()
```

Comment:
Shouldn't we add the event handler before creating the node to avoid possible race conditions?

Reply:
make sense and done :)